### ISSUE TITLE
Add more extensive unit tests for fast / approximate math methods

### ISSUE THREAD

**Issue by tskisner** [hpc4cmb/toast#289]
We need more detailed tests around extreme values for the internal fast trig functions.

**Comment by tskisner** [comment]
In addition to this, we should expand the test case for the internal healpix projection functions.  This can be done from python to allow comparison to healpy.  Although the current compiled unit tests compare ang2pix results around the sphere, they do not compare results for small epsilons around extrema of the projection (pole, phi=0, etc).  Since this compiled code will be used by larger kernels (for example on a GPU), it is important to establish more rigorous consistency with healpy.  I'll work on this after #288 is merged.

**Comment by mreineck** [comment]
Out of curiosity: do you have benchmark results on the fast vector functions? My feeling is that, at least for big array sizes, memory bandwidth will negate most of the speed advantages ... especially for cheap functions like `sqrt`.
Also, recent gcc versions do a very good vectorization of mathematical intrinsics with the right compiler flags (-ffast-math -O3 -march=...).

**Comment by tskisner** [comment]
@mreineck , I agree that in the era of modern compilers and architectures with wide SIMD units and deeper memory hierarchies, all of these Taylor approximation methods should be revisited for their usefulness or else removed.  This code was written 10 or more years ago at a time when we were expanding the detector pointing every iteration of a conjugate gradient solver.  Things like atan2() became the dominant cost of the entire map-making job, and these approximations reduced the runtime greatly compared to the functions in libm.  Currently we expand the pointing once and save it.  However in the future we will likely move back to on-the-fly pointing calculation for some architectures and some pipelines.  For example, a filter-and-bin mapmaking does not need to re-use the pointing, so can compute it once when needed and then discard it.  For iterative solvers on a CPU, we may still want to save the pointing if it will fit in RAM.  Our CUDA tests show that doing the pointing expansion on a modern (Volta) GPU on-demand is subdominant compared to an expensive mapmaking technique like a GLS solver using FFTs.

Since this issue was opened, we switched all of our internal pointing code to use the plain libm functions.  Also the expanded healpy unit test comparisons were added in #296.  So I think the original topic of this issue is closed.  However, feel free to open another issue for discussion on pointing topics, etc.
